base/genericmemory.jl

```diff
@@ -422,3 +422,72 @@ function replaceindex_atomic!(
         @_boundscheck,
     )
 end
+
+"""
+     allocatedinline(::Type{T}) :: Bool
+
+Returns whether an object of type `T` is able to be inline allocated inside of enclosing structs or arrays,
+of if it must be allocated as a reference. Generally, types which are immutable and whose instances have a
+finite, statically determinable size can be allocated inline, whereas everything else is allocated using
+reference pointers.
+
+# Examples
+```jldoctest
+julia> Base.allocatedinline(Int)
+true
+
+julia> Base.allocatedinline(Union{Int, Float64})
+true
+
+julia> Base.allocatedinline(Real)
+true
+
+julia> Base.allocatedinline(Complex)
```

Comment:
The `Real` and `Complex` feel like they serve the same purpose so only one is maybe enough?